vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java

```diff
@@ -555,14 +554,28 @@ public GridDataView<T> getGenericDataView() {
      * @return the created hierarchy column
      */
     public Column<T> addHierarchyColumn(ValueProvider<T, ?> valueProvider) {
-        Column<T> column = addColumn(TemplateRenderer
-                .<T> of("<vaadin-grid-tree-toggle "
-                        + "leaf='[[!item.children]]' expanded='{{expanded}}' level='[[level]]'>[[item.name]]"
-                        + "</vaadin-grid-tree-toggle>")
+
+        // The click listener needs to check if the event gets canceled (by
+        // vaadin-grid-tree-toggle) and only invoke the callback if it does.
+        // vaadin-grid-tree-toggle will cancel the event if the user clicks on
+        // a non-focusable element inside the toggle.
+        Column<T> column = addColumn(LitRenderer.<T> of(
+                "<vaadin-grid-tree-toggle @click=${onClick} .leaf=${!item.children} .expanded=${model.expanded} .level=${model.level}>"
+                        + "${item.name}</vaadin-grid-tree-toggle>")
                 .withProperty("children",
                         item -> getDataCommunicator().hasChildren(item))
                 .withProperty("name",
-                        value -> String.valueOf(valueProvider.apply(value))));
+                        value -> String.valueOf(valueProvider.apply(value)))
+                .withFunction("onClick", item -> {
+                    if (getDataCommunicator().hasChildren(item)) {
```

Comment:
Should we add this condition to `HierarchyComponentRenderer` as well?
https://github.com/vaadin/flow-components/blob/158b50b10a3acd9a5b104d738c4015d284e0be39/vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/HierarchyColumnComponentRenderer.java#L27-L33

Reply:
In `HierarchyComponentRenderer`s case, with the conditional click listener, the [check is done client-side](https://github.com/vaadin/web-components/blob/d6f871f2e694aa08748481d32ae9c7f6ba2e6b20/packages/grid/src/vaadin-grid-tree-toggle.js#L192)